src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp

```diff
@@ -166,6 +167,118 @@ tnsr::ijj<DataType, SpatialDim, Frame> deriv_spatial_metric(
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_unit_normal,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi) noexcept {
+  auto dt_shift =
+    make_with_value<tnsr::I<DataType, SpatialDim, Frame>>(get(lapse), 0.);
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    for (size_t j = 0; j < SpatialDim; ++j) {
+      for (size_t a = 0; a < SpatialDim + 1; ++a) {
+        // first term
+        dt_shift.get(i) -= get(lapse) * pi.get(j+1, a)
+            * spacetime_unit_normal.get(a) * inverse_spatial_metric.get(i, j);
+        // second term
+        for (size_t k = 0; k < SpatialDim; ++k) {
+          dt_shift.get(i) += shift.get(j)
+              * spacetime_unit_normal.get(a) * phi.get(j, k+1, a)
+              * inverse_spatial_metric.get(i, k);
+        }
+      }
+    }
+    dt_shift.get(i) *=  get(lapse);
+  }
+  return dt_shift;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::i<DataType, SpatialDim, Frame> time_deriv_of_lower_shift(
+    const Scalar<DataType>& lapse,
+    const tnsr::I<DataType, SpatialDim, Frame>& shift,
+    const tnsr::ii<DataType, SpatialDim, Frame>& spatial_metric,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::ii<DataType, SpatialDim, Frame>& dt_spatial_metric,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_unit_normal,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi) noexcept {
+  auto dt_lower_shift =
+    make_with_value<tnsr::i<DataType, SpatialDim, Frame>>(get(lapse), 0.);
+  // get \partial_0 N^j
+  const auto dt_shift = time_deriv_of_shift(lapse, shift,
+    inverse_spatial_metric, spacetime_unit_normal, phi, pi);
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    for (size_t j = 0; j < SpatialDim; ++j) {
+      dt_lower_shift.get(i) += spatial_metric.get(i, j) * dt_shift.get(j)
+        + shift.get(j) * dt_spatial_metric.get(i, j);
+    }
+  }
+  return dt_lower_shift;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::a<DataType, SpatialDim, Frame> spacetime_deriv_of_norm_of_shift(
+    const Scalar<DataType>& lapse,
+    const tnsr::I<DataType, SpatialDim, Frame>& shift,
+    const tnsr::ii<DataType, SpatialDim, Frame>& spatial_metric,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::ii<DataType, SpatialDim, Frame>& dt_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_unit_normal,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi) noexcept {
+  auto d4_norm_of_shift =
+    make_with_value<tnsr::a<DataType, SpatialDim, Frame>>(get(lapse), 0.);
+  // compute intermediate quantities
+  const auto lower_shift = raise_or_lower_index(shift, spatial_metric);
```

Comment:
Could you use a `Variables` for the temporaries and call the `not_null` versions of the functions? See, for example, https://github.com/sxs-collaboration/spectre/blob/5e9c717596069f53a1e349f17e3cfaaf5c9b1ee5/src/Evolution/Systems/GrMhd/ValenciaDivClean/Characteristics.cpp#L109

Reply:
also pushed